### DISCUSSION TITLE
How to chain fzf commands?

### DISCUSSION THREAD

**Discussion by elenapan** [ibhagwan/fzf-lua#1237] [Q&A] [answered]
For example, I would like to first select a directory from `project_list` and then start a file search in that directory.

How would I go about implementing this?

The following snippet does not work as expected.

```lua
local project_list = {
    vim.fn.expand("~")..'/.config',
    vim.fn.expand("~")..'/bin',
    vim.fn.expand("~")..'/notes'
}
vim.keymap.set('n', '<leader>pp', function()
    local fzf = require("fzf-lua")
    local choice = fzf.fzf_exec(project_list)
    fzf.files({ cwd = choice })
end)
```

**Comment by ibhagwan** [comment, marked as answer]
Hi @elenapan, what an honor for you to be interested in my project, to this day I still use a color scheme for my terminal from one of your customizations :)

As for your question, fzf-lua is async and runs in a lua coroutine so you can’t chain the commands the way you tried, you need to create a custom action, see:
https://github.com/ibhagwan/fzf-lua/wiki/Advanced#fzf-exec-acts

```lua
fzf.fzf_exec(project_list, {
  actions = {
    ['default'] = {
      fn = function(sel, opts)
        fzf.files({ cwd = sel[1] })
      end,
      -- we use `noclose` to reuse the existing floating windows
      -- it will work regardless but you’ll see the windows “flash”
      noclose = 1,
    },
    -- anything is possible with some effort and lua
    ['ctrl-y'] = function(selected, opts)
      print("selected item:", selected[1])
    end
  }
})
```

I personally use something similar in my config which switches the `cwd` and also lets me go back to the previous active one:
![IMG_2153](https://github.com/ibhagwan/fzf-lua/assets/59988195/ca118ebd-9df3-4074-8063-19dfc2f6d29a)

Code can be found here:
https://github.com/ibhagwan/nvim-lua/blob/d6a3dd1de64414b48231d6498b901f8c5a8bc9d3/lua/plugins/fzf-lua/cmds.lua#L66-L109

**Reply by elenapan** [reply]
Thank you for pointing me to the right direction. Your snippet worked perfectly.

Also, to return the kind words, you have done an amazing job, fzf-lua is the one plugin I cannot live without :)

**Reply by AlejandroSanchez90** [reply]
> Hi @elenapan, what an honor for you to be interested in my project, to this day I still use a color scheme for my terminal from one of your customizations :)
> 
> As for your question, fzf-lua is async and runs in a lua coroutine so you can’t chain the commands the way you tried, you need to create a custom action, see: https://github.com/ibhagwan/fzf-lua/wiki/Advanced#fzf-exec-acts
> 
> ```lua
> fzf.fzf_exec(project_list, {
>   actions = {
>     ['default'] = {
>       fn = function(sel, opts)
>         fzf.files({ cwd = sel[1] })
>       end,
>       -- we use `noclose` to reuse the existing floating windows
>       -- it will work regardless but you’ll see the windows “flash”
>       noclose = 1,
>     },
>     -- anything is possible with some effort and lua
>     ['ctrl-y'] = function(selected, opts)
>       print("selected item:", selected[1])
>     end
>   }
> })
> ```
> 
> I personally use something similar in my config which switches the `cwd` and also lets me go back to the previous active one: ![IMG_2153](https://private-user-images.githubusercontent.com/59988195/337614705-ca118ebd-9df3-4074-8063-19dfc2f6d29a.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyMzM5NDIsIm5iZiI6MTcyNjIzMzY0MiwicGF0aCI6Ii81OTk4ODE5NS8zMzc2MTQ3MDUtY2ExMThlYmQtOWRmMy00MDc0LTgwNjMtMTlkZmMyZjZkMjlhLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxM1QxMzIwNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNWYzOGM2MDkxNzNiYmIzYmRjZjlhZmViNmM2MGQ2NjcwMDBmMDMwZWE3NzA1NDFlOTllN2FlY2JlMDcxYzQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tuJFECKv7Umtq9OA5CLWTwJwLTOkZmZobL1JqOARNvA)
> 
> Code can be found here: https://github.com/ibhagwan/nvim-lua/blob/d6a3dd1de64414b48231d6498b901f8c5a8bc9d3/lua/plugins/fzf-lua/cmds.lua#L66-L109

hey, im trying to use your function to set the cwd with a picker, i do see the picker but its empty, im using yout exact code from https://github.com/ibhagwan/nvim-lua/blob/d6a3dd1de64414b48231d6498b901f8c5a8bc9d3/lua/plugins/fzf-lua/cmds.lua#L66-L109

any idea?

**Reply by ibhagwan** [reply]
I don’t know, you’d have to debug it, use old school peint debugging to see if the callback is called and with what values.
